cidata.py

```diff
@@ -0,0 +1,431 @@
+        metadata:
+          description: data structure that contains the machine's metadata
+          type: object
+          required: true
+          properties:
+            instance_id:
+              type: string
+              required: true
+              description: libvirt domain name for VM. Typically composed from
+                           numerical project ID and VM ID as follows:
+                           `<project_id>_<domain_id>`.
+            network:
+              type: object
+              required: true
+              description: the VM's network configuration
+              properties:
+                nameservers:
+                  description: the VM's DNS configuration
+                  type: object
+                  required: true
+                  properties:
+                    addresses:
+                      description: list of name server IP addresses
+                      type: array
+                      required: true
+                      items:
+                        description: a name server IP address such as `8.8.8.8`. At least one must be specified.
+                        type: string
+                        required: true
+                    search:
+                      type: array
+                      description: the machine's search domains for unqualified host names
+                      required: false
+                      items:
+                        description: a search domain to qualify unqualified
+                                     host names with, such as `cloudcix.com`
+                        required: false
+                        type: string
+                interfaces:
+                  type: object
+                  required: true
+                  description: The VM's network interface configuration
+                  properties:
+                    mac_address:
+                      description: The interface's MAC address (colon separated
+                                   bytes, lower case)
+                      type: string
+                      required: true
+                    addresses:
+                      description: The interface's IP addresses. At least one is required.
+                      type: array
+                      required: true
+                      items:
+                        description: an IP address with subnet mask specified
+                                     in CIDR notation as understood by ip(8), e.g.
+                                     `10.0.5.221/24`
+                        type: string
+                        required: true
+                    routes:
+                      description: routes to create for this particular interface. While optional, setting at
+                                   least a default route is highly recommended.
+                      type: object
+                      required: false
+                      properties:
+                        to:
```

Comment:
```suggestion
                        destination:
```

It was agreed last week that `destination` and `gateway` would be used as the keys in the routes dictionary.

Reply:
We could do that, but then we'd either have to change cloud-init again to add a translation to `via` and `to` or modify the primitive to do this translation. The reason behind these two terms is that cloud-init uses the netplan data structure for its internal representation of network configuration which can be passed through directly:

https://github.com/CloudCIX/cloud-init/blob/daacae77e07e2df24106a3d49013b1a308835c73/cloudinit/sources/DataSourceCloudCIX.py#L138

If we have different key names they need to be translated somewhere.

Reply:
It would be managed in the jinja template e.g. This was my thought process based on the spreadsheet:

The format required would be defined in the jinja template and the variable values would populate the template with the values in the sent contract. 

```
"instance_id": "{{ domain_path }}",
"network": 
    "nameservers": {"addresses": ["{{ dns }}"], "search": ["cloudcix.com"]},
    "interfaces": [
{% for interface in interfaces %}
        {
            "mac_address": "{{ interface['mac_address'] }}",
            "routes": [
{% for route in routes %}
               {"to": "{{ route['destaination'] }}", "via": "{{ route['gateway'] }}"}],
{% endear %}
            "addresses": {{ interface['ip_addresses'],
        },
{% endfor %}
    ]
}
```